lib/torrent.js

```diff
@@ -1348,71 +1348,78 @@ class Torrent extends EventEmitter {
   _rechoke () {
     if (!this.ready) return
 
-    if (this._rechokeOptimisticTime > 0) this._rechokeOptimisticTime -= 1
-    else this._rechokeOptimisticWire = null
+    // wires in increasing order of quality (pop() gives next best peer)
+    const wireStack =
+      this.wires
+        .map(wire => ({ wire, random: Math.random() })) // insert a random seed for randomizing the sort
+        .sort((objA, objB) => {
+          const wireA = objA.wire
+          const wireB = objB.wire
+
+          // prefer peers that send us data faster
+          if (wireA.downloadSpeed() !== wireB.downloadSpeed()) {
+            return wireA.downloadSpeed() - wireB.downloadSpeed()
+          }
 
-    const peers = []
+          // then prefer peers that can download data from us faster
+          if (wireA.uploadSpeed() !== wireB.uploadSpeed()) {
+            return wireA.uploadSpeed() - wireB.uploadSpeed()
+          }
 
-    this.wires.forEach(wire => {
-      if (!wire.isSeeder && wire !== this._rechokeOptimisticWire) {
-        peers.push({
-          wire,
-          downloadSpeed: wire.downloadSpeed(),
-          uploadSpeed: wire.uploadSpeed(),
-          salt: Math.random(),
-          isChoked: true
-        })
-      }
-    })
+          // then prefer already unchoked peers (to minimize fibrillation)
+          if (wireA.amChoking !== wireB.amChoking) {
+            return wireA.amChoking ? -1 : 1 // choking < unchoked
+          }
 
-    peers.sort(rechokeSort)
+          // otherwise random order
+          return objA.random - objB.random
+        })
+        .map(obj => obj.wire) // return array of wires (remove random seed)
 
-    let unchokeInterested = 0
-    let i = 0
-    for (; i < peers.length && unchokeInterested < this._rechokeNumSlots; ++i) {
-      peers[i].isChoked = false
-      if (peers[i].wire.peerInterested) unchokeInterested += 1
+    if (this._rechokeOptimisticTime <= 0) {
+      // clear old optimistic peer, so it can be rechoked normally and then replaced
+      this._rechokeOptimisticWire = null
+    } else {
+      this._rechokeOptimisticTime -= 1
     }
 
-    // Optimistically unchoke a peer
-    if (!this._rechokeOptimisticWire && i < peers.length && this._rechokeNumSlots) {
-      const candidates = peers.slice(i).filter(peer => peer.wire.peerInterested)
-      const optimistic = candidates[randomInt(candidates.length)]
+    let numInterestedUnchoked = 0
+    while (numInterestedUnchoked < this._rechokeNumSlots - 1) { // leave one slot open for optimistic unchoking
+      const wire = wireStack.pop() // next best quality peer
```

Comment:
I would check if `wireStack` has any elements left. Or do something like this:

```suggestion
      const wire = wireStack.pop() // next best quality peer
      if (!wire) break
```

Reply:
nice catch